pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -123,69 +145,99 @@ protected List<Object[]> buildJoinedRows(TransferableBlock leftBlock)
       case ANTI:
         return buildJoinedDataBlockAnti(leftBlock);
       default: { // INNER, LEFT, RIGHT, FULL
-        return buildJoinedDataBlockDefault(leftBlock);
+        if (_rightTable.isKeysUnique()) {
```

Comment:
I've manually validated the changes here and the logic looks good but I think it would be nice to have more unit tests covering this at a more granular level.

Reply:
Actually on second thought, I think the tests in `HashJoinOperatorTest` are already covering many of these scenarios.